#include <string.h>
#include <stdio.h>
#include <stdlib.h>

void encodeNuc(char *filename){
    /*
     * This function reads the nucleobase sequence from a text file
     * and generates a new text file containing the corresponding binary sequence.
     * (COMMENT OUT PRINT LATER)
     */
    FILE * fileptr;
    fileptr = fopen(filename, "r");
    FILE * newfile;
    char *name = filename;
    char new_name[100] = "b";
    strcat(new_name, name);
    newfile = fopen(new_name, "w");

    if (fileptr == NULL){
        exit(0);
    }

    char base = getc(fileptr);
    char c = 'C';
    char a = 'A';
    char g = 'G';
    char t = 'T';


    while (base != EOF){
        if (base == a){
            fputs("00", newfile);
        }
        else if (base == c){
            fputs("01", newfile);
        }
        else if (base == g){
            fputs("10", newfile);
        }
        else if (base == t){
            fputs("11", newfile);
        }
        base = getc(fileptr);
    }
    fclose(fileptr);
    fclose(newfile);
}

void decodeBin (char *filename){
    /*
     * read the binary sequence from a text file
     * read every two characters
     * generates a new text file containing the nucleobase sequence
     */
    FILE * fileptr;
    fileptr = fopen(filename, "r");
    FILE * newfile;
    char *name = filename;
    char new_name[100] = "n";
    strcat(new_name, name);
    newfile = fopen(new_name, "w");

    if (fileptr == NULL){
        exit(0);
    }

    char n1 = getc(fileptr);
    char n2 = getc(fileptr);
    char zero = '0';

    while (n2 != EOF){
        if (n1 == zero){
            if (n2 == zero){
                fputs("A", newfile);
            }
            else{
                fputs("C", newfile);
            }
        }
        else{
            if (n2 == zero){
                fputs("G", newfile);
            }
            else{
                fputs("T", newfile);
            }
        }
        n1 = getc(fileptr);
        n2 = getc(fileptr);
    }
    fclose(fileptr);
    fclose(newfile);
}

// let's read a nucleobase sequence first
void findProtein(char *filename, int checkPos, int proteinInfo[]){
    /*
     * read binary sequence
     * convert into nucleobase
     * open the new file generated by decodeBIN
     * given a starting point
     * identifies the first protein found
     * set proteinInfo[] = {0,0}
     * set proteinInfo[0] = nucleobase position of the start of the codon
     * counter needed
     * NOTE: count from one not zero--> count+1 is the answer
     * set proteinInfo[1] = animo acid length of the protein
     */


    FILE * newfile;
    decodeBin(filename);
    int position = checkPos;
    int length = 0;
    int start = 0;
    char *name = filename;
    char new_name[100] = "n";
    strcat(new_name, name);
    newfile = fopen(new_name, "r");

    if (checkPos == 0){
        printf("Exit -- position starts counting at 1");
        exit(0);
    }

    char c = 'C';
    char a = 'A';
    char g = 'G';
    char t = 'T';
    fseek(newfile, (position-1) * sizeof(char), SEEK_SET);
    char n1 = getc(newfile);
    char n2 = getc(newfile);
    char n3 = getc(newfile);
    while (n3 != EOF){
        if (n1 == a && n2 == t && n3 == g){ //start
            length++;
            start = 1;
            n1 = getc(newfile);
            n2 = getc(newfile);
            n3 = getc(newfile);
            continue;
        }
        else if (start == 1 && ((n1 == t && n2 == a && n3 == a) || (n1 == t && n2 == a && n3 == g)|| (n1 == t && n2 == g && n3 == a))){
            length++;
            break;
        }
        else if (start == 0){
            position++;
            n1 = n2;
            n2 = n3;
            n3 = getc(newfile);
        }
        else{
            length++;
            n1 = getc(newfile);
            n2 = getc(newfile);
            n3 = getc(newfile);
        }
    }
    if (start == 0){
        // there is no DNA
        proteinInfo[0] = 0;
        proteinInfo[1] = length;
    }
    else{
        proteinInfo[0] = position;
        proteinInfo[1] = length;
    }
    fclose(newfile);
}

long int count_char(char*fn){
    FILE * fileptr;
    long int count = 0;
    fileptr = fopen(fn, "r");
    char chr = getc(fileptr);
    while(chr != EOF){
        count = count + 1;
        chr = getc(fileptr);
    }
    fclose(fileptr);
    return count;
}

void proteinReport(char * filename){
    /*
     * This function reads a binary sequence from a text file and generates a new text file containing information
     * about all the protein regions in the sequence.
     */
    int position = 1;
    int proteinInfo[] = {0,0};
    FILE * fileptr;
    FILE * newfile;
    fileptr = fopen(filename, "r");
    char *name = filename;
    char new_name[100] = "r";
    strcat(new_name, name);
    newfile = fopen(new_name, "w");
    long int number = count_char(filename);
    while(position<=number){
        findProtein(filename, position, proteinInfo);
        if (proteinInfo[0] != 0 && proteinInfo[1] != 0){
            fprintf(newfile, "%d,%d\n", proteinInfo[0], proteinInfo[1]);
            position = proteinInfo[0] + 1;
        }
        else{
            break;
        }
    }
    fclose(fileptr);
    fclose(newfile);
}

void isolateProtein(char *filename, int proteinInfo[]){
    /*
     * This function reads a binary sequence from a text file
     * and generates a new text file containing the amino acid translations of a specified protein.
     */

    if (proteinInfo[0] == 0){
        printf("No protein");
        exit(0);
    }

    FILE * fileptr;
    fileptr = fopen(filename, "r");
    FILE * newfile;

    // file to be written
    char *name = filename;
    char new_name[100] = "p";
    strcat(new_name, name);
    newfile = fopen(new_name, "w");

    // get information
    int length = proteinInfo[1];// number of sets

    // for loop
    char n1 = fseek(fileptr, (proteinInfo[0] - 1) * 2 * sizeof(char), SEEK_SET);
    n1 = getc(fileptr);
    char n2 = getc(fileptr);
    char n3 = getc(fileptr);
    char n4 = getc(fileptr);
    char n5 = getc(fileptr);
    char n6 = getc(fileptr);
    char zero = '0';
    char one = '1';
    for (int i = 0; i < length; i++){
        if(n1 == one && n2 == one){ //T
            if(n3 == one && n4 == one){//T
                if(n6 == one){
                    fputs("F", newfile);
                }
                else{
                    fputs("L", newfile);
                }
            }
            else if(n3 == zero && n4 == one){//C
                fputs("S", newfile);
            }
            else if(n3 == one && n4 == zero){//G
                if(n6 == one){
                    fputs("C", newfile);
                }
                else if (n5 == one && n6==zero){
                    fputs("W", newfile);
                }
                else{
                    fputs("*", newfile);
                    break;
                }
            }
            else{ //A
                if(n6 == one){
                    fputs("Y", newfile);
                }
                else{
                    fputs("*", newfile);
                    break;
                }
            }
        }
        else if(n1 == zero && n2 == one){//C
            if (n3 == one && n4 == one){//T
                fputs("L", newfile);
            }
            else if (n3 == zero && n4 == one){//C
                fputs("P", newfile);
            }
            else if(n3 == zero && n4 == zero && n6 == one){//A
                fputs("H", newfile);
            }
            else if(n3 == zero && n4 == zero && n6 == zero){//A
                fputs("Q", newfile);
            }
            else if(n3 == one && n4 == zero){
                fputs("R", newfile);
            }
        }
        else if(n1 == one && n2 == zero){//G
            if (n3 == one && n4 == one){
                fputs("V", newfile);
            }
            else if (n3 == zero && n4 == one){
                fputs("A", newfile);
            }
            else if (n3 == zero && n4 == zero){
                if (n6 == one){
                    fputs("D", newfile);
                }
                else{
                    fputs("E", newfile);
                }
            }
            else{
                fputs("G", newfile);
            }
        }
        else{ //A
            if (n3 == one && n4 == one){
                if (n5 == one && n6 == zero){
                    fputs("M", newfile);
                }
                else{
                    fputs("V", newfile);
                }
            }
            else if (n3 == zero && n4 == one){
                fputs("T", newfile);
            }
            else if (n3 == zero && n4 == zero){
                if (n6 == one){
                    fputs("N", newfile);
                }
                else{
                    fputs("K", newfile);
                }
            }
            else{
                if (n6 == one){
                    fputs("S", newfile);
                }
                else{
                    fputs("R", newfile);
                }
            }
        }
    n1 = getc(fileptr);
    n2 = getc(fileptr);
    n3 = getc(fileptr);
    n4 = getc(fileptr);
    n5 = getc(fileptr);
    n6 = getc(fileptr);
    }
    fclose(newfile);
    fclose(fileptr);
}

int genMutant(char *filename, int mutation[]){
    /*
     * This function reads a binary sequence from a text file
     * and generates a new text file containing a copy of
     * the sequence with a specifed point mutation.
     */
    decodeBin(filename);
    FILE * buffer;
    FILE * newfile;
    FILE * old = fopen(filename,"r");

    char *bname = filename;
    char b_name[100] = "n";
    strcat(b_name, bname);
    buffer = fopen(b_name, "r");

    char *name = filename;
    char new_name[100] = "m";
    strcat(new_name, name);
    newfile = fopen(new_name, "w");

    int position = mutation[0]; //based on nucleobase eg. position = 2
    int type = mutation[1];
    int base = mutation[2];

    int counter = 1;
    char mutate;
    while ( (mutate = getc(buffer)) != EOF ){
        if (counter == position){
            break;
        }
        else{
            counter++;
        }
    }
    fclose(buffer);
    int mutate_num;
    char a = 'A';
    char c = 'C';
    char g = 'G';
    char ch;

    if (mutate == a) mutate_num = 0;
    else if (mutate == c) mutate_num = 1;
    else if (mutate == g) mutate_num = 2;
    else mutate_num = 3;

    if (type == 0){
        if (base == mutate_num){
            int count = 0;
            while( ( ch = getc(old) ) != EOF ){
                if (count == ((position-1) * 2)) {
                    break;
                }
                else{
                    fputc(ch, newfile);
                    count++;
                }
            }
            getc(old);
            while( ( ch = getc(old) ) != EOF )
                fputc(ch, newfile);
            fclose(old);
            fclose(newfile);
            return 0;
        }
        else{
            while( ( ch = getc(old) ) != EOF )
                fputc(ch, newfile);
            fclose(old);
            fclose(newfile);
            return 1;
        }
    }
    else if (type == 1){ // nucleobase is inserted (complete)
        int count = 0;
        while ( (ch = getc(old)) != EOF){
            if (count == (position * 2 - 1)){
                if (base == 0){
                    fputc(ch, newfile);
                    fputc('0', newfile);
                    fputc('0', newfile);
                    count++;
                }
                else if (base == 1){
                    fputc(ch, newfile);
                    fputc('0', newfile);
                    fputc('1', newfile);
                    count++;
                }
                else if (base == 2){
                    fputc(ch, newfile);
                    fputc('1', newfile);
                    fputc('0', newfile);
                    count++;
                }
                else{
                    fputc(ch, newfile);
                    fputc('1', newfile);
                    fputc('1', newfile);
                    count++;
                }
            }
            else{
                fputc(ch, newfile);
                count++;
            }
        }
        fclose(old);
        fclose(newfile);
        return 0;
    }
    else{ // replaced with a different nucleobase
        int count = 0;
        int success = 0;
        while( ( ch = getc(old) ) != EOF ){
            if (count == ((position-1) * 2)){
                if (base == 0){
                    fputc('0', newfile);
                    fputc('0', newfile);
                    getc(old);
                    count++;
                    success = 1;
                }
                else if (base == 1){
                    fputc('0', newfile);
                    fputc('1', newfile);
                    getc(old);
                    count++;
                    success = 1;
                }
                else if (base == 2){
                    fputc('1', newfile);
                    fputc('0', newfile);
                    getc(old);
                    count++;
                    success = 1;
                }
                else{
                    fputc('1', newfile);
                    fputc('1', newfile);
                    getc(old);
                    count++;
                    success = 1;
                }
            }
            else{
                fputc(ch, newfile);
                count++;
            }
        }
        if (success == 1){
            fclose(old);
            fclose(newfile);
            return 0;
        }
        else{
            fclose(old);
            fclose(newfile);
            return 1;
        }
    }
}
/*
int checkMutant(char * oriFilename, char * mutFilename){
    FILE * ori = fopen(oriFilename, "r");
    FILE * mut = fopen(mutFilename, "r");
    FILE * new;

    // file to be written
    char *name = mutFilename;
    char new_name[100] = "c";
    strcat(new_name, name);
    new = fopen(new_name, "w");

    // length of the file
    fseek(ori, 0L, SEEK_END);
    long int size_ori = ftell(ori);
    fseek(mut, 0L, SEEK_END);
    long int size_mut = ftell(mut);
    rewind(ori);
    rewind(mut);

    char ch_ori;
    char ch_mut;
    long int location;

    //convert into nucleobase

    if (size_mut == size_ori){ // must be substitution
        while ((ch_ori = getc(ori) != EOF)){
            ch_mut = getc(mut);
            if (ch_ori == ch_mut){
                continue;
            }
            else{ // there is a mutant
                location = ftell(ori);
                int mutate_num;
                if (ch_ori == 'A') mutate_num = 0;
                else if (ch_ori == 'C') mutate_num = 1;
                else if (ch_ori == 'G') mutate_num = 2;
                else mutate_num = 3;
                fprintf(new, "%ld,%d,%d", location, 0, mutate_num);

            }
        }
    }
    else{
        printf("HI");
    }
    // returning the distance
    int result_ori;
    int result_mut;
    int protein[] = {0,0};
    findProtein(ori, 1, protein);
    result_ori = protein[0];
    findProtein(mut, 1, protein);
    result_mut = protein[0];
    fclose(ori);
    fclose(new);
    fclose(mut);
    return abs(result_mut - result_ori);
}
*/
int main(){
    encodeNuc("fullSLV.txt");
    printf("encoded \n");
    //decodeBin("bpartialSLV.txt");
    //printf("decoded \n");
    int proteinInfo[] = {0,0};
    findProtein("bfullSLV.txt", 1, proteinInfo);
    proteinReport("bfullSLV.txt");
    proteinInfo[0] = 120;
    proteinInfo[1] = 276;
    isolateProtein("bfullSLV.txt", proteinInfo);
    proteinInfo[0] = 33;
    proteinInfo[1] = 10;
    isolateProtein("bfullSLV.txt", proteinInfo);
    //printf("\n");
    int mutation[] = {1,0,3};
    genMutant("bfullSLV.txt", mutation);
    printf("ran sucessfully!");
    return 0;
}

